database/src/ops/block.rs

```diff
@@ -1,89 +1,492 @@
+    // }
 
-/// TODO
-pub fn get_block_hash() {
-    todo!()
-}
+    let cumulative_rct_outs = get_rct_num_outputs(tables.rct_outputs())?;
 
-/// TODO
-pub fn get_block_height() {
-    todo!()
-}
+    // Block Info.
+    tables.block_infos_mut().put(
+        &block.height,
+        &BlockInfo {
+            timestamp: block.block.header.timestamp,
+            total_generated_coins: block.generated_coins,
+            cumulative_difficulty: block.cumulative_difficulty,
+            block_hash: block.block_hash,
+            cumulative_rct_outs,
+            // INVARIANT: #[cfg] @ lib.rs asserts `usize == u64`
+            weight: block.weight as u64,
+            long_term_weight: block.long_term_weight as u64,
+        },
+    )?;
 
-/// TODO
-pub fn get_block_weight() {
-    todo!()
-}
+    // Block blobs.
+    tables.block_blobs_mut().put(
+        &block.height,
+        bytemuck::TransparentWrapper::wrap_ref(&block.block_blob),
+    )?;
 
-/// TODO
-pub fn get_block_already_generated_coins() {
-    todo!()
-}
+    // Block heights.
+    tables
+        .block_heights_mut()
+        .put(&block.block_hash, &block.height)?;
 
-/// TODO
-pub fn get_block_long_term_weight() {
-    todo!()
-}
+    // Cast height to `u32` for storage.
+    // Panic (should never happen) instead of allowing DB corruption.
+    // <https://github.com/Cuprate/cuprate/pull/102#discussion_r1560020991>
+    let height = u32::try_from(block.height).expect("height was > u32::MAX");
+
+    // Transaction & Outputs.
+    for tx in &block.txs {
+        add_tx(tx, tables)?;
+        let tx: &Transaction = &tx.tx;
+
+        // ^
+        // Everything above is adding tx data.
+        // Everything below is for adding input/output data.
+        // v
+
+        // Is this a miner transaction?
+        // Which table we add the output data to depends on this.
+        // <https://github.com/monero-project/monero/blob/eac1b86bb2818ac552457380c9dd421fb8935e5b/src/blockchain_db/blockchain_db.cpp#L212-L216>
+        let mut miner_tx = false;
+
+        for inputs in &tx.prefix.inputs {
+            match inputs {
+                // Key images.
+                Input::ToKey { key_image, .. } => {
+                    add_key_image(key_image.compress().as_bytes(), tables.key_images_mut())?;
+                }
+                // This is a miner transaction, set it for later use.
+                Input::Gen(_) => miner_tx = true,
+            }
+        }
+
+        // Output bit flags.
+        // Set to a non-zero bit value if the unlock time is non-zero.
+        // TODO: use bitflags.
+        let output_flags = match tx.prefix.timelock {
+            Timelock::None => OutputFlags::NONE,
+            Timelock::Block(_) | Timelock::Time(_) => OutputFlags::NON_ZERO_UNLOCK_TIME,
+        };
 
-/// TODO
-pub fn get_block_timestamp() {
-    todo!()
+        // Output data.
+        for (i, output) in tx.prefix.outputs.iter().enumerate() {
```

Comment:
it doesn't look like the [`tx outputs`](https://github.com/Cuprate/cuprate/tree/db-schema/database#tx-outputs) table is being filled?

Reply:
Just took a look at tables.rs. There are no defined tables that link TxID to amount idxs...
Documentation error ? or table mistake ?

Reply:
Oops, that table wasn't even implemented... added table and updated `add_block()`